### PR TITLE
Alerting: Persist silence state immediately on Create/Delete 

### PR DESCRIPTION
**Pull request by JacobsonMT**
**What is this feature?**

This feature makes silence Create/Delete in GMA update the snapshot in kvstore immediately instead of waiting for the
next maintenance run. This is to prevent silences from being lost when a new Alertmanager is started before the state has persisted. This can happen, for example, in a rolling deployment scenario when HA is not enabled.

**Why do we need this feature?**

Currently, there is a bug that can happen with specific timings of silence creation and pod restart.

Conditions:
- HA is not enabled.
- Pod is restarted with something like rolling deploys, where the new instance is started before the old instance is stopped.
- A silence is created and then the pod is restarted before the next maintenance run.

With the above conditions, the newly created silences will eventually be lost.

This happens because silence state is only persisted to db:
- Periodically during maintenance (default 15m)
- On shutdown.

While silence state is loaded from db (keeping in mind, no gossiping without HA):
- On startup.

This causes the newly deployed instance to read a state from the db that does not include any of these new silences and it will never re-load the database. From the new instances point of view, the new silences do not exist. In this interim, alerts that should be silenced can slip through if they are attempted to be sent by this AM.

Eventually, the old instance will shutdown and write the silence state to the db but the new instance will not read this state. Finally, once the new instance's maintenance runs, the db silence state will be overwritten with the incomplete set of silence, permanently losing the new silences.

**Who is this feature for?**

Users of Grafana-Managed Alerting.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/alerting-squad/issues/761

**Special notes for your reviewer:**

- Will update `grafana/alerting` dependency to real ref once [PR](https://github.com/grafana/alerting/pull/169) is merged.
- Eventually, we will likely want to treat silences as top-level Resources so they can be incorporated in future API/CRD plans. This means storing them individually instead of as a blob so we can guarantee consistency on writes in an HA environment. I chose to bring us slightly closer to that goal but without changing anything too drastically yet as there are too many unknowns right now.
- The interface is likely to be a point of contention and I welcome suggestions. I can't think of any clearly better options that don't increase the scope of work here to a large degree:
  - It would have been preferable to use the existing `GettableSilence` methods, but they do not contain `ExpiresAt` necessary for the state snapshot.
  - `SilenceState(context.Context) (alertingNotify.SilenceState, error)` could have instead returned the string encoding at the expense of type safety. The gain of a bit of performance on a low-volume endpoint seems to not be worth it though.
  - We currently have to make a call to the AM implementation to get the silence state. For local AM, this is not a problem. For remote AM, this means an extra network call. This is obviously not great, but I don't think there is a clear improvement to be made here without going full CRD/async reconciliation.
- We can put this behind a FF if that's desired but seeing as this is a bugfix at heart, I would prefer not to.
- For remote alertmanager maintainers: The way the remote AM currently reads state from the kvstore but never saves it back seems wrong and I believe is likely not working as expected for primary remote AM. Either way, fixing the logic was out of scope for this PR, I implemented `SilenceState` according to what made sense to me.

### CODE REVIEW

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 1]
```diff
@@ -431,6 +435,74 @@ func (moa *MultiOrgAlertmanager) AlertmanagerFor(orgID int64) (Alertmanager, err
...
+		return "", ErrAlertmanagerNotReady
+	}
+
+	// Need to create the silence in the AM first to get the silence ID.
+	silenceID, err := orgAM.CreateSilence(ctx, ps)
+	if err != nil {
+		return "", err
+	}
+
+	return silenceID, moa.updateSilenceState(ctx, orgAM, orgID)
```
Comment (rwwiv) [inline pkg/services/ngalert/notifier/multiorg_alertmanager.go:459]: Do we want to return an error if the silence has been created but couldn't be persisted? Persistence is technically not necessary for the silence to be applied, just to ensure it's not lost across restarts.

**Review by rwwiv — APPROVED** [review]
LGTM, my comment doesn't need to be addressed to make it in.

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go (hunk 1 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go]: Fixed: f4aa99a228a74c47c39871dc545124ba3ffcbf44

**Pull request merged**